### PR TITLE
[RLlib; docs] Docs do-over (new API stack): RLModules vol 03; Actual rst rewrite and upgrade.

### PR DESCRIPTION
**Pull request by sven1977**
Docs do-over (new API stack): RLModules vol 03
* Actual rst rewrite and upgrade.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/rllib-examples.rst [hunk 1]
```diff
@@ -200,7 +200,7 @@ Hierarchical Training
    learning over flat approaches by demonstrating significantly improved learning performance in challenging, goal-oriented tasks.
 
 
-Inference (of Models/Policies)
+Inference (of models/policies)
```
Comment (angelinalg) [inline doc/source/rllib/rllib-examples.rst:203]:
```suggestion
Inference of models or policies
```

Comment (angelinalg) [reply doc/source/rllib/rllib-examples.rst:203]: https://developers.google.com/style/parentheses

File: doc/source/rllib/rllib-examples.rst [hunk 2]
```diff
@@ -200,7 +200,7 @@ Hierarchical Training
...
 ++++++++++++++++++++++++++++++
 
 - `Policy Inference after Training <https://github.com/ray-project/ray/blob/master/rllib/examples/inference/policy_inference_after_training.py>`__:
```
Comment (angelinalg) [inline doc/source/rllib/rllib-examples.rst:206]:
```suggestion
- `Policy inference after training <https://github.com/ray-project/ray/blob/master/rllib/examples/inference/policy_inference_after_training.py>`__:
```

File: doc/source/rllib/rllib-examples.rst [hunk 3]
```diff
@@ -234,7 +234,7 @@ Metrics
    performance- and environment indicators beyond the standard RLlib metrics.
 
 
-Multi-Agent RL
+Multi-agent RL
 ++++++++++++++
 
 - `Custom Heuristic Policy <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/custom_heuristic_policy.py>`__:
```
Comment (angelinalg) [inline doc/source/rllib/rllib-examples.rst:240]:
```suggestion
- `Custom heuristic policy <https://github.com/ray-project/ray/blob/master/rllib/examples/multi_agent/custom_heuristic_policy.py>`__:
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 4]
```diff
@@ -2,169 +2,505 @@
...
+.. _rlmodule-guide:
 
-.. |pytorch| image:: images/pytorch.png
-    :class: inline-figure
-    :width: 16
+RL Modules
+==========
 
+The :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` class in RLlib's new API stack allows you to write custom
+models in RLlib, including highly complex multi-network setups, often found in multi-agent- or model-based algorithms.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:11]:
```suggestion
models in RLlib, including highly complex multi-network setups, often found in multi-agent or model-based algorithms.
```

**Review by angelinalg — COMMENTED** [review]
I still have many files to review, but I will release these comments first.

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 5]
```diff
@@ -403,6 +395,99 @@ to the new API stack's :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`:
...
+            greedy_actions = my_dist.to_deterministic().sample()
+            return {Columns.ACTIONS: greedy_actions}
+
+
+Policy.compute_log_likelihoods
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Implement your custom RLModule's :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule._forward_train` method and
+return the `Columns.ACTION_LOGP` key together with the corresponding action log probs in order to pass this information
+on to your loss function(s) (which are called right after `forward_train()`. The loss logic can then access
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:488]: Your correction is correct, thanks!

File: doc/source/rllib/package_ref/rl_modules.rst [hunk 6]
```diff
@@ -23,81 +23,81 @@ Single Agent
...
     RLModule.as_multi_rl_module
 
 
 Forward methods
 +++++++++++++++
 
+Use the following three forward methods when you use your RLModule from inside other classes
+and components. However, do NOT override them and leave them as-is in your custom subclasses.
+For defining your own forward behavior, override the private methods ``_forward`` (generic forward behavior for
+all phases) or - for more granularity - ``_forward_exploration``, ``_forward_inference``, and ``_forward_train``.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/package_ref/rl_modules.rst:78]: What I wanted to say is that users should really not touch these methods at all in their subclasses (but override the private counterparts instead). I'll try to clarify this better ...

File: doc/source/rllib/rllib-rlmodule.rst [hunk 7]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+.. _rllib-rl-module-w-custom-action-dists:
 
-            class MyPPORLModule(PPORLModule):
+Custom action distributions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-                def __init__(self, config: RLModuleConfig):
-                    super().__init__(config)
-                    ...
+In the preceding examples, you relied on your RLModule to use the correct action distribution with the computed
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:623]:
```suggestion
The preceding examples rely on the RLModule uses the correct action distribution with the computed
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 8]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+``ACTION_DIST_INPUTS`` returned by your forward methods. RLlib picks a default distribution class based on
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:624]:
```suggestion
``ACTION_DIST_INPUTS`` returned by the forward methods. RLlib picks a default distribution class based on
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 9]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+the action space, which is :py:class:`~ray.rllib.models.torch.torch_distributions.TorchCategorical` for ``Discrete`` action spaces
+and :py:class:`~ray.rllib.models.torch.torch_distributions.TorchDiagGaussian` for ``Box`` action spaces.
 
-            # Pass in the custom RL Module class to the spec
-            algo_config = algo_config.rl_module(
-                rl_module_spec=RLModuleSpec(module_class=MyPPORLModule)
-            )
+If you want to use a different distribution class (and return parameters for its constructor from your forward methods),
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:628]:
```suggestion
To use a different distribution class and return parameters for its constructor from your forward methods,
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 10]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+override the following methods in your :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` implementation:
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:629]:
```suggestion
override the following methods in the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` implementation:
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 11]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
-        class to create your custom encoder instead of the default one. We do this in the following example.
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.get_inference_action_dist_cls`
+- :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.get_exploration_action_dist_cls`
+- and :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.get_train_action_dist_cls`
 
-        .. literalinclude:: ../../../rllib/examples/rl_modules/classes/mobilenet_rlm.py
-                :language: python
-                :start-after: __sphinx_doc_begin__
-                :end-before: __sphinx_doc_end__
+`Commonly used distribution implementations can be found here <https://github.com/ray-project/ray/blob/master/rllib/models/torch/torch_distributions.py>`__.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:635]:
```suggestion
See `torch_distributions.py <https://github.com/ray-project/ray/blob/master/rllib/models/torch/torch_distributions.py>`__ for common distribution implementations.
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 12]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
 
+Implement a deterministic counterpart of your distribution class through the
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:637]:
```suggestion
Implement a deterministic counterpart of the distribution class through the
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 13]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+:py:meth:`~ray.rllib.models.distributions.Distribution.to_deterministic` method.
+You can choose to compute deterministic actions, by creating the deterministic counterpart.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:639]:
```suggestion
You can choose to compute deterministic actions by creating the deterministic counterpart.
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 14]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
-        This approach is useful mostly if you want your custom component to integrate with the decision trees that the Catalogs represent.
-        The following use cases are examples of what may require you to extend the Catalogs:
+Implementing custom MultiRLModules
+----------------------------------
 
-            - Choosing a custom model only for a certain observation space.
-            - Using a custom action distribution in multiple distinct Algorithms.
-            - Reusing your custom component in many distinct RL Modules.
+For multi-module setups, RLlib provides the :py:class:`~ray.rllib.core.rl_module.multi_rl_module.MultiRLModule` class,
+whose default implementation is a dictionary of individual :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` objects,
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:646]:
```suggestion
whose default implementation is a dictionary of individual :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` objects.
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 15]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+one for each submodule and identified by a ``ModuleID``.
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:647]:
```suggestion
Each submodule has an RLModule identified by a ``ModuleID``.
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 16]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
 
-        For instance, to adapt existing ``PPORLModules`` for a custom graph observation space not supported by RLlib out-of-the-box,
-        extend the :py:class:`~ray.rllib.core.models.catalog.Catalog` class used to create the ``PPORLModule``
-        and override the method responsible for returning the encoder component to ensure that your custom encoder replaces the default one initially provided by RLlib.
+The base-class :py:class:`~ray.rllib.core.rl_module.multi_rl_module.MultiRLModule` implementation works for most of those
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:649]:
```suggestion
The base-class :py:class:`~ray.rllib.core.rl_module.multi_rl_module.MultiRLModule` implementation works for most of the
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 17]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+use cases that need to define independent neural networks . However, for any complex, multi-network or multi-agent use cases, where agents share one or more neural networks,
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:650]:
```suggestion
use cases that need to define independent neural networks. However, for any complex, multi-network or multi-agent use case, where agents share one or more neural networks,
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 18]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+you should inherit from this class and override the default implementation.
 
-        .. code-block:: python
+The following code snippets create a custom multi-agent RL module with two simple "policy head" modules, which
+share the same encoder (the third network in the MultiRLModule). The encoder receives the raw observations from the env
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:654]:
```suggestion
share the same encoder, the third network in the MultiRLModule. The encoder receives the raw observations from the env
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 19]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+        Within your MultiRLModule, you need to have two policy sub-RLModules. They may be of the same
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:674]:
```suggestion
        Within a MultiRLModule, you need to have two policy sub-RLModules. They may be of the same
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 20]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
+            :end-before: __sphinx_doc_policy_end__
+        .. literalinclude:: ../../../rllib/examples/rl_modules/classes/vpg_using_shared_encoder_rlm.py
+            :language: python
+            :start-after: __sphinx_doc_policy_2_begin__
+            :end-before: __sphinx_doc_policy_2_end__
 
+    .. tab-item:: Shared encoder RLModule
 
-What your customization could have looked like before:
+        Finally, your shared encoder RLModule should look similar to this:
```
Comment (angelinalg) [inline doc/source/rllib/rllib-rlmodule.rst:688]:
```suggestion
        Finally, the shared encoder RLModule should look similar to this:
```

**Review by angelinalg — COMMENTED** [review]
Another installment of comments. Still at around line 700 of`doc/source/rllib/rllib-rlmodule.rst`.

File: doc/source/rllib/rllib-rlmodule.rst [hunk 21]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
                     return {
-                        # RLlib will:
-                        # - Generate distribution from these parameters.
-                        # - Convert distribution to a deterministic equivalent.
-                        # - "sample" from the deterministic distribution.
-                        "action_dist_inputs": ...
+                        # RLlib:
+                        # - Generates distribution from ACTION_DIST_INPUTS parameters.
+                        # - Converts distribution to a deterministic equivalent.
+                        # - Samples from the (deterministic) distribution.
```
Comment (sven1977) [inline doc/source/rllib/rllib-rlmodule.rst:545]:
```suggestion
                        # - Samples from the deterministic distribution.
```

File: doc/source/rllib/package_ref/rl_modules.rst [hunk 22]
```diff
@@ -5,15 +5,15 @@
 
 .. _rlmodule-reference-docs:
 
-RLModule API
-============
+RLModule APIs
+=============
 
 
 RL Module specifications and configurations
```
Comment (sven1977) [inline doc/source/rllib/package_ref/rl_modules.rst:12]:
```suggestion
RLModule specifications and configurations
```

File: doc/source/rllib/package_ref/rl_modules.rst [hunk 23]
```diff
@@ -5,15 +5,15 @@
...
 -------------------------------------------
 
-Single Agent
-++++++++++++
+Single RLModule
++++++++++++++++
```
Comment (sven1977) [inline doc/source/rllib/package_ref/rl_modules.rst:16]:
```suggestion
Single RLModuleSpec
+++++++++++++++++++
```

File: doc/source/rllib/package_ref/rl_modules.rst [hunk 24]
```diff
@@ -23,81 +23,83 @@ Single Agent
 
     RLModuleSpec
     RLModuleSpec.build
-    RLModuleSpec.get_rl_module_config
 
-RLModule Configuration
-+++++++++++++++++++++++
+MultiRLModule
++++++++++++++
```
Comment (sven1977) [inline doc/source/rllib/package_ref/rl_modules.rst:28]:
```suggestion
MultiRLModuleSpec
+++++++++++++++++
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 25]
```diff
@@ -2,169 +2,505 @@
...
+            )
+
 
-If you return the "actions" key:
+Forward methods
+~~~~~~~~~~~~~~~~~~~
 
-- RLlib will use the actions provided thereunder as-is.
-- If you also returned the "action_dist_inputs" key: RLlib will also create a :py:class:`~ray.rllib.models.distributions.Distribution` object from the distribution parameters under that key and - in the case of :py:meth:`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` - compute action probs and logp values from the given actions automatically.
+After that, implement the forward computation logic. You can either define a generic forward logic by overriding the
```
Comment (sven1977) [inline doc/source/rllib/rllib-rlmodule.rst:467]:
```suggestion
Implementing the forward computation logic, you can either define a generic forward behavior by overriding the
```

File: doc/source/rllib/rllib-rlmodule.rst [hunk 26]
```diff
@@ -173,359 +509,296 @@ Commonly used distribution implementations can be found under ``ray.rllib.models
...
                     return {
-                        # RLlib will:
-                        # - Generate distribution from these parameters.
-                        # - "sample" from the (stochastic) distribution.
-                        # - Compute action probs/logs automatically using the sampled
-                        #   actions and the generated distribution object.
-                        "action_dist_inputs": ...
+                        # RLlib:
+                        # - Generates distribution from ACTION_DIST_INPUTS parameters.
+                        # - Samples from the (stochastic) distribution.
```
Comment (sven1977) [inline doc/source/rllib/rllib-rlmodule.rst:554]:
```suggestion
                        # - Samples from the stochastic distribution.
```

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 27]
```diff
@@ -22,29 +22,21 @@ RLlib classes and code to RLlib's new API stack.
...
 
 
-Change your AlgorithmConfig
----------------------------
+Check your AlgorithmConfig
+--------------------------
 
-RLlib turns on the new API stack by default for all RLlib algorithms. To deactivate it, use the `api_stack()` method
-in your `AlgorithmConfig` object like so:
+RLlib turns on the new API stack by default for all RLlib algorithms.
```
Comment (simonsays1980) [inline doc/source/rllib/new-api-stack-migration-guide.rst:28]: Didn't I change this already in here: https://github.com/ray-project/ray/pull/49304 ?

File: doc/source/rllib/new-api-stack-migration-guide.rst [hunk 28]
```diff
@@ -22,29 +22,21 @@ RLlib classes and code to RLlib's new API stack.
...
 
-.. testcode::
+.. note::
+    To **deactivate** the new API stack and switch back to the old one, use the
```
Comment (simonsays1980) [inline doc/source/rllib/new-api-stack-migration-guide.rst:31]: We should leave here a comment for when we see the old stack being deprectaed and not supported anymore. I see way too many users on the discussion forums to go with the old stack, even when starting their journey. I always mention that the old stack will be deprecated soon, but without a date.

File: rllib/algorithms/marwil/marwil.py [hunk 29]
```diff
@@ -265,10 +264,7 @@ def get_default_rl_module_spec(self) -> RLModuleSpecType:
                 PPOTorchRLModule,
             )
 
-            return RLModuleSpec(
-                module_class=PPOTorchRLModule,
-                catalog_class=MARWILCatalog,
-            )
+            return RLModuleSpec(module_class=PPOTorchRLModule)
```
Comment (simonsays1980) [inline rllib/algorithms/marwil/marwil.py:267]: `DefaultPPOTorchRLModule`?

File: rllib/algorithms/marwil/marwil_learner.py [hunk 30]
```diff
@@ -41,3 +42,10 @@ def remove_module(
         )
         # In case of BC (beta==0.0 and this property never being used),
         self.moving_avg_sqd_adv_norms_per_module.pop(module_id, None)
+
+    @classmethod
+    @override(Learner)
+    def rl_module_required_apis(cls) -> list[type]:
```
Comment (simonsays1980) [inline rllib/algorithms/marwil/marwil_learner.py:48]: Nice!

File: rllib/core/learner/learner.py [hunk 31]
```diff
@@ -1565,6 +1549,36 @@ def _make_module(self) -> MultiRLModule:
 
         return module
 
+    def rl_module_is_compatible(self, module: RLModule) -> bool:
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:1552]: Yes, this is what we can use now where almost everything is defined by interfaces.

File: rllib/examples/rl_modules/classes/vpg_rlm.py [hunk 32]
```diff
@@ -1,9 +1,7 @@
+import torch
```
Comment (simonsays1980) [inline rllib/examples/rl_modules/classes/vpg_rlm.py:1]: No `try_import_torch` anymopre needed?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Very nice changes!

File: rllib/examples/rl_modules/classes/vpg_rlm.py (hunk 32 above)
Comment (sven1977) [reply rllib/examples/rl_modules/classes/vpg_rlm.py:1]: Good point, but I think we can leave the bare-metal `import torch` everywhere that's not in the "common" import path (and thus doesn't break anything for non-torch users).

File: rllib/algorithms/marwil/marwil.py (hunk 29 above)
Comment (sven1977) [reply rllib/algorithms/marwil/marwil.py:267]: Great catch. These are covered, though, by backward compatible imports in the old py-files.

Comment (sven1977) [reply rllib/algorithms/marwil/marwil.py:267]: changed

File: doc/source/rllib/new-api-stack-migration-guide.rst (hunk 28 above)
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:31]: Yup! Let's change the "new stack" header everywhere describing that they should ONLY use the new stack from here on (unless they have very good reasons to still use the old one). ...

File: doc/source/rllib/new-api-stack-migration-guide.rst (hunk 27 above)
Comment (sven1977) [reply doc/source/rllib/new-api-stack-migration-guide.rst:28]: Yeah, but I removed the "to deactivate it (<- the new API stack), ...". I think we shouldn't describe to users how to deactivate the new API stack on the migration guide site. It's too confusing. Basically, we are saying now: "It's already on, don't worry! But you can double-check your config by doing xyz ..."

**Pull request merged**